StringIndexOutOfBoundsException from FlutterLogEntryParser

## What happened

Every time I open a Flutter project I see this error. I saw there was a similar report from 2016 but the issue was closed, so I've opened a new issue for visibility.

## Version information

IntelliJ IDEA `2019.1.3` • Flutter plugin `io.flutter 37.0.3` • Dart plugin `191.7830`

Flutter 1.7.8+hotfix.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision b712a172f9 (7 days ago) • 2019-07-09 13:14:38 -0700
Engine • revision 54ad777fd2
Tools • Dart 2.4.0

## Exception

null

```
java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor83.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.execution.process.ProcessHandler$2.invoke(ProcessHandler.java:214)
	at com.sun.proxy.$Proxy23.onTextAvailable(Unknown Source)
	at com.intellij.execution.process.ProcessHandler.notifyTextAvailable(ProcessHandler.java:188)
	at com.intellij.execution.process.BaseOSProcessHandler$SimpleOutputReader.onTextAvailable(BaseOSProcessHandler.java:179)
	at com.intellij.util.io.BaseOutputReader.sendText(BaseOutputReader.java:211)
	at com.intellij.util.io.BaseOutputReader.processInput(BaseOutputReader.java:195)
	at com.intellij.util.io.BaseOutputReader.readAvailableBlocking(BaseOutputReader.java:149)
	at com.intellij.util.io.BaseDataReader.readAvailable(BaseDataReader.java:77)
	at com.intellij.util.io.BaseDataReader.doRun(BaseDataReader.java:155)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:224)
	at com.intellij.util.io.BaseDataReader.lambda$start$0(BaseDataReader.java:61)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1960)
	at io.flutter.logging.text.LineParser.parse(LineParser.java:112)
	at io.flutter.logging.FlutterLogEntryParser$LineHandler.parseLineStyle(FlutterLogEntryParser.java:289)
	at io.flutter.logging.FlutterLogEntryParser$LineHandler.parseLineInfo(FlutterLogEntryParser.java:307)
	at io.flutter.logging.FlutterLogEntryParser$LineHandler.parseEntry(FlutterLogEntryParser.java:312)
	at io.flutter.logging.FlutterLogEntryParser.parseDaemonEvent(FlutterLogEntryParser.java:348)
	at io.flutter.logging.FlutterLogEntryParser.parseDaemonEvent(FlutterLogEntryParser.java:272)
	at io.flutter.logging.FlutterLog$1.onTextAvailable(FlutterLog.java:168)
	... 19 more
```

--- a/src/io/flutter/console/FlutterConsoleFilter.java
+++ b/src/io/flutter/console/FlutterConsoleFilter.java
@@ -112,6 +112,9 @@ public class FlutterConsoleFilter implements Filter {
 
     int lineNumber = 0;
     String pathPart = line.trim();
+    VirtualFile file = null;
+    int lineStart = -1;
+    int highlightLength = 0;
 
     // Check for, e.g.,
     //   * "Launching lib/main.dart"
@@ -120,6 +123,11 @@ public class FlutterConsoleFilter implements Filter {
       final String[] parts = pathPart.split(" ");
       if (parts.length > 1) {
         pathPart = parts[1];
+        file = fileAtPath(pathPart);
+        if (file != null) {
+          lineStart = line.indexOf(pathPart);
+          highlightLength = pathPart.length();
+        }
       }
     }
 
@@ -127,8 +135,6 @@ public class FlutterConsoleFilter implements Filter {
     //    * "  • MyApp.xzzzz (lib/main.dart:6)"
     //    * "  • _MyHomePageState._incrementCounter (lib/main.dart:49)"
     final String[] parts = pathPart.split(" ");
-    int lineStart = -1;
-    int highlightLength = 0;
     for (String part : parts) {
       // "(lib/main.dart:49)"
       if (part.startsWith("(") && part.endsWith(")")) {
@@ -167,7 +173,9 @@ public class FlutterConsoleFilter implements Filter {
       }
     }
 
-    final VirtualFile file = fileAtPath(pathPart);
+    if (file == null) {
+      file = fileAtPath(pathPart);
+    }
     if (file != null) {
       // "open ios/Runner.xcworkspace"
       final boolean openAsExternalFile = FlutterUtils.isXcodeFileName(pathPart);

--- a/testSrc/unit/io/flutter/console/FlutterConsoleFilterTest.java
+++ b/testSrc/unit/io/flutter/console/FlutterConsoleFilterTest.java
@@ -42,4 +42,11 @@ public class FlutterConsoleFilterTest {
     final Filter.Result link = new FlutterConsoleFilter(fixture.getModule()).applyFilter(line, 659);
     assertNotNull(link);
   }
+
+  @Test
+  public void checkLaunchingLink() {
+    String line = "Launching test/widget_test.dart on Android SDK built for x86 in debug mode...\n";
+    Filter.Result link = new FlutterConsoleFilter(fixture.getModule()).applyFilter(line, line.length());
+    assertNotNull(link);
+  }
 }

--- a/testSrc/unit/io/flutter/console/FlutterConsoleFilterTest.java
+++ b/testSrc/unit/io/flutter/console/FlutterConsoleFilterTest.java
@@ -5,30 +5,30 @@
  */
 package io.flutter.console;
 
+import static org.junit.Assert.assertNotNull;
+
 import com.intellij.execution.filters.Filter;
 import com.intellij.openapi.roots.ModuleRootModificationUtil;
 import com.intellij.openapi.vfs.VirtualFile;
 import io.flutter.testing.ProjectFixture;
 import io.flutter.testing.TestDir;
 import io.flutter.testing.Testing;
-import org.junit.Before;
-import org.junit.Rule;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
 import org.junit.Test;
 
-import static org.junit.Assert.assertNotNull;
-
 public class FlutterConsoleFilterTest {
-  @Rule
-  public final ProjectFixture fixture = Testing.makeCodeInsightModule();
+  @ClassRule
+  public static final ProjectFixture fixture = Testing.makeEmptyProject();
 
-  @Rule
-  public final TestDir tmp = new TestDir();
+  @ClassRule
+  public static final TestDir tmp = new TestDir();
 
-  VirtualFile contentRoot;
-  String appDir;
+  static VirtualFile contentRoot;
+  static String appDir;
 
-  @Before
-  public void setUp() throws Exception {
+  @BeforeClass
+  public static void setUp() throws Exception {
     contentRoot = tmp.ensureDir("root");
     appDir = tmp.ensureDir("root/test").getPath();
     tmp.writeFile("root/test/widget_test.dart", "");

--- a/testSrc/unit/io/flutter/testing/Testing.java
+++ b/testSrc/unit/io/flutter/testing/Testing.java
@@ -14,10 +14,9 @@ import com.intellij.testFramework.builders.EmptyModuleFixtureBuilder;
 import com.intellij.testFramework.fixtures.IdeaProjectTestFixture;
 import com.intellij.testFramework.fixtures.IdeaTestFixtureFactory;
 import com.intellij.testFramework.fixtures.TestFixtureBuilder;
-
-import javax.swing.*;
 import java.lang.reflect.InvocationTargetException;
 import java.util.concurrent.atomic.AtomicReference;
+import javax.swing.SwingUtilities;
 
 /**
  * Test utilities.

